STACK-UP-3/warriorz-backend
# #171784097 add trip request notifications

#### What this PR does?
 -Delivers trip request notifications' resources to the project.

#### Description of the task to be completed?
- adding a notification controller to the controllers.
- adding socket connection for real-time email notifications.
- adding notificationHandle(notification creator and email sender).
- adding swagger.
- adding tests and validations for the feature

#### How should this be tested?
 - First, you clone the repo on your machine.
 - Open the folder in your text editor,  *preferably   `visual studio code`*
 - Now open your terminal, *for `visual studio code`    just click on    `terminal`* 
 - Now run `npm install`
 - Then  run `npm run start:dev`,
- First `signup`, verify email, and `sign in`.
- After now create a trip request.
- To view all notification run route: ` localhost:5000/api/v1/notifications?page={page}&limit={limit}`
- To view a specific notification use route:  ` localhost:5000/api/v1/notifications/:notification_id`
> NB: both `notification_id` and `page,limit` are of type integer.

#### Integrating it to your feature.
- First, import the `notificationHandle` file into your controller.
- Then create objects `emailContent` and `notificationData` where the attributes of the objects are as follows:
 > **emailContent** = {
        subject: _email subject_ ,
        userMessage: _message that will be sent to the user in the email._,
        managerMessage:  _message that will be sent to the user in the email._,
      };
      **notificationData** = {
        trip_request_id: _integer_,
        user_id: _integer_,
        line_manager_id: _integer_,
        subject: _string_,
        description: _integer_,
      };

- Then pass in the notification handle like this 
 ``` 
await notificationHandle(notificationData,userId,emailContent,req);
Where userId = req.userData.id.
 ```

#### What are the relevant pivotal tracker stories?
  [#171784097](https://www.pivotaltracker.com/story/show/171784097)

#### Screenshots?
##### Notification Creation:
1. Realtime notification:

![image](https://user-images.githubusercontent.com/52471122/80857074-816e0e80-8c4f-11ea-81a0-3379cb5b1ce4.png)

2. Email sent to the user upon notification creation:

![image](https://user-images.githubusercontent.com/52471122/80830683-af206c80-8be9-11ea-9b48-a9cdc08fdb34.png)

3. Email sent to the manager upon notification creation:

![image](https://user-images.githubusercontent.com/52471122/80857825-4c64ba80-8c55-11ea-9f55-96bcbf0271a7.png)

##### Viewing Notifications:

1. View all notifications:

![image](https://user-images.githubusercontent.com/52471122/80925170-f8a9bc80-8d8d-11ea-9296-2cfcf1cb395e.png)

2. View a specific notification:

![image](https://user-images.githubusercontent.com/52471122/80925182-0c552300-8d8e-11ea-9415-7d169ff84058.png)
